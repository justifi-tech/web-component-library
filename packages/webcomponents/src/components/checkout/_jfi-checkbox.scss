input[type="checkbox"] {
  cursor: pointer;
}

.checkbox-label {
  cursor: auto;
}