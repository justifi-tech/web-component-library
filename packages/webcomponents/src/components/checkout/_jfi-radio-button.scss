input[type="radio"],
label {
  cursor: pointer;
}

input[type="radio"] {
  background-color: var(--jfi-radio-button-background-color);
  border-color: var(--jfi-radio-button-border-color);

  &:checked,
  &:checked:focus {
    background-color: var(--jfi-radio-button-background-color-selected);
    border-color: var(--jfi-radio-button-border-color-selected);
  }

  &:focus {
    border-color: var(--jfi-radio-button-border-color-focus);
    box-shadow: var(--jfi-radio-button-box-shadow-focus);
  }
}