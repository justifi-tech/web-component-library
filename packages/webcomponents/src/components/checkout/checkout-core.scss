@import 'reboot';
@import 'grid';
@import 'buttons';
@import 'spinners';
@import 'list-group';
@import './jfi-header';

:host {
  @include host-base-styles;
  display: block;
}

justifi-payment-method-form {
  display: block;
  margin-left: calc(var(--jfi-layout-padding) * -1);
  margin-right: calc(var(--jfi-layout-padding) * -1);
}

.jfi-checkout-core {
  color: var(--jfi-body-color);
}

*::part(label) {
  color: var(--jfi-form-label-color);
  font-family: var(--jfi-form-label-font-family);
  font-weight: var(--jfi-form-label-font-weight);
  font-size: var(--jfi-form-label-font-size);
  margin: var(--jfi-form-label-margin);
}

*::part(input) {
  font-family: var(--jfi-layout-font-family);
  background-color: var(--jfi-form-control-background-color);
  border-color: var(--jfi-form-control-border-color);
  border-width: var(--jfi-form-control-border-width);
  border-bottom-width: var(--jfi-form-control-border-bottom-width);
  border-left-width: var(--jfi-form-control-border-left-width);
  border-right-width: var(--jfi-form-control-border-right-width);
  border-top-width: var(--jfi-form-control-border-top-width);
  border-radius: var(--jfi-form-control-border-radius);
  border-style: var(--jfi-form-control-border-style);
  box-shadow: var(--jfi-form-control-box-shadow);
  color: var(--jfi-form-control-color);
  font-size: var(--jfi-form-control-font-size);
  font-weight: var(--jfi-form-control-font-weight);
  line-height: var(--jfi-form-control-line-height);
  margin: var(--jfi-form-control-margin);
  padding: var(--jfi-form-control-padding);

  &:focus {
    color: var(--jfi-form-control-color-focus);
    border-color: var(--jfi-form-control-border-color-focus);
    box-shadow: var(--jfi-form-control-box-shadow-focus);
  }
}

*::part(input-invalid) {
  border-color: var(--jfi-form-control-border-color-error);
  box-shadow: var(--jfi-form-control-box-shadow-error);

  &:focus {
    border-color: var(--jfi-form-control-border-color-error);
    box-shadow: var(--jfi-form-control-box-shadow-error-focus);
  }
}

*::part(invalid-feedback) {
  color: var(--jfi-error-message-color);
  font-size: var(--jfi-error-message-font-size);
  margin: var(--jfi-error-message-margin);
  line-height: normal;
}

.jfi-submit-button {
  font-family: var(--jfi-layout-font-family);
  color: var(--jfi-submit-button-color);
  background-color: var(--jfi-submit-button-background-color);
  border-color: var(--jfi-submit-button-border-color);
  padding: var(--jfi-submit-button-padding);
  font-size: var(--jfi-submit-button-font-size);
  border-radius: var(--jfi-submit-button-border-radius);
  width: var(--jfi-submit-button-width);
  box-shadow: var(--jfi-submit-button-box-shadow);
  line-height: var(--jfi-submit-button-line-height);
  text-transform: var(--jfi-submit-button-text-transform);
}

.jfi-submit-button:hover {
  color: var(--jfi-submit-button-color-hover);
  background-color: var(--jfi-submit-button-background-color-hover);
  border-color: var(--jfi-submit-button-border-color-hover);
}

.jfi-submit-button:focus {
  color: var(--jfi-submit-button-color-focus);
  background-color: var(--jfi-submit-button-background-color-focus);
  border-color: var(--jfi-submit-button-border-color-focus);
}

// Extra selector for specificity
.btn.jfi-submit-button:active {
  color: var(--jfi-submit-button-color-active);
  background-color: var(--jfi-submit-button-background-color-active);
  border-color: var(--jfi-submit-button-border-color-active);
}

.btn.jfi-submit-button-loading {
  min-width: 75px;
  color: var(--jfi-submit-button-color-loading);
  background-color: var(--jfi-submit-button-background-color-loading);
  border-color: var(--jfi-submit-button-border-color-loading);
}

.btn.jfi-submit-button-loading .spinner-border {
  border-top-color: var(--jfi-submit-button-color-loading);
  border-bottom-color: var(--jfi-submit-button-color-loading);
  border-left-color: var(--jfi-submit-button-color-loading);
}