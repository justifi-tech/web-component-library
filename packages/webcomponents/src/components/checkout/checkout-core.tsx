import { Component, h, Prop, State, Event, EventEmitter, Host, Listen, Method } from '@stencil/core';
import { CreatePaymentMethodResponse } from '../payment-method-form/payment-method-responses';
import { extractComputedFontsToLoad } from '../../utils/utils';
import { PaymentMethodTypes } from '../../api';
import { config } from '../../../config';

@Component({
  tag: 'justifi-checkout-core',
  styleUrl: 'checkout-core.scss',
  shadow: true,
})
export class CheckoutCore {
  /**
 * URL for the rendered iFrame. End-users need not use this.
 */
  @Prop({ mutable: true }) iframeOrigin?: string = config.iframeOrigin; @Prop() authToken: string;
  @Prop() getCheckout: Function;
  @Prop() checkoutId: string;

  @State() hasLoadedFonts: boolean = false;
  @State() isLoading: boolean = false;
  @State() selectedPaymentMethodType: PaymentMethodTypes = PaymentMethodTypes.card;
  @State() checkout: any = {};
  @State() serverError: boolean = false;
  @State() errorMessage: string = '';

  @Event() submitted: EventEmitter<CreatePaymentMethodResponse>;

  private paymentMethodFormRef?: HTMLJustifiPaymentMethodFormElement;
  private billingFormRef?: HTMLJustifiBillingFormElement;

  componentWillLoad() {
    if (this.getCheckout) {
      this.fetchData();
    }
  }

  connectedCallback() {
    if (!this.hasLoadedFonts) {
      this.loadFontsOnParent();
      this.hasLoadedFonts = true;
    }
  }

  fetchData(): void {
    this.isLoading = true;

    this.getCheckout({
      onSuccess: ({ checkout }) => {
        this.checkout = checkout;
        this.isLoading = false;
      },
      onError: (errorMessage) => {
        this.errorMessage = errorMessage;
        this.isLoading = false;
      },
    });
  };

  @Listen('paymentMethodSelected')
  paymentMethodSelectedHandler(event: CustomEvent) {
    const paymentMethodType: PaymentMethodTypes = event.detail;
    this.selectedPaymentMethodType = paymentMethodType;
  }

  @Method()
  async loadFontsOnParent() {

    const parent = document.body;
    const fontsToLoad = extractComputedFontsToLoad();
    if (!parent || !fontsToLoad) {
      return null;
    }

    // This approach is needed to load the font in a parent of the component
    const fonts = document.createElement('link');
    fonts.rel = 'stylesheet';
    fonts.href = `https://fonts.googleapis.com/css2?family=${fontsToLoad}&display=swap`;

    parent.append(fonts);
  }

  async submit(event) {
    event.preventDefault();
    if (!this.paymentMethodFormRef || !this.billingFormRef) return;

    const billingFormValidation = await this.billingFormRef.validate();
    const paymentMethodFormValidation = await this.paymentMethodFormRef.validate();

    if (!billingFormValidation.isValid || !paymentMethodFormValidation.isValid) return;

    this.isLoading = true;

    try {
      const billingFormFieldValues = await this.billingFormRef.getValues();
      const paymentMethodData = { ...billingFormFieldValues };
      const clientId = this.checkout.payment_client_id;
      const tokenizeResponse = await this.paymentMethodFormRef.tokenize(clientId, paymentMethodData);
      if (tokenizeResponse.error) {
        console.error(`An error occured submitting the form: ${tokenizeResponse.error.message}`);
      }
      this.submitted.emit(tokenizeResponse);
    } catch (error) {
      console.error(`An error occured submitting the form: ${error}`);
    } finally {
      this.isLoading = false;
    }
  }

  private loadingSpinner = (
    <div class="spinner-border spinner-border-sm" role="status">
      <span class="visually-hidden">Loading...</span>
    </div>
  );

  render() {
    return (
      <Host>
        <form class="row gy-3">
          <justifi-payment-method-type-selector
            show-credit-card={this.checkout.payment_settings?.credit_card_payments}
            show-ach={this.checkout.payment_settings?.ach_payments}
            selected-payment-method-type={this.selectedPaymentMethodType}
          />
          <div class="col-12">
            <justifi-payment-method-form
              payment-method-form-type={this.selectedPaymentMethodType}
              iframe-origin={this.iframeOrigin}
              ref={el => {
                if (el) { this.paymentMethodFormRef = el; }
              }}
            />
          </div>
          <div class="col-12">
            <justifi-billing-form
              legend="Billing Info"
              ref={el => {
                if (el) { this.billingFormRef = el; }
              }}
            />
          </div>
          <slot name='insurance' />
          <div class="col-12">
            <button
              type="submit"
              onClick={event => this.submit(event)}
              disabled={this.isLoading}
              class={`btn btn-primary jfi-submit-button ${this.isLoading ? 'jfi-submit-button-loading' : ''}`}
            >
              {this.isLoading ? this.loadingSpinner : 'Submit'}
            </button>
          </div>
        </form>
      </Host>
    );
  }
}