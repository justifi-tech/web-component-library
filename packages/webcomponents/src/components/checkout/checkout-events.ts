import { ICheckoutCompleteResponse } from "../../api/Checkout";

export type CheckoutSubmitEvent = {
  response: ICheckoutCompleteResponse;
}