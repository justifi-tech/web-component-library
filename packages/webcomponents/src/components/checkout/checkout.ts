export interface ICheckout {
  id: string;
  type: string;
  page_info: null;
  data: {
    payment_intent_id: string;
    account_id: string;
    platform_account_id: string;
    payment_amount: string;
    payment_client_id: string;
    payment_description: string;
    payment_methods: {
      id: string;
      status: string;
      invalid_reason: null;
      name: string;
      brand: string;
      acct_last_four: string;
      month: string;
      year: string;
      address_line1_check: string;
      address_postal_code_check: string;
      bin_details: null;
    }[];
    payment_method_group_id: string;
    payment_settings: {
      ach_payments: boolean;
      bnpl_payments: boolean;
      credit_card_payments: boolean;
    };
    bnpl?: {
      provider: string;
      provider_client_id: string;
      provider_mode: string;
      provider_checkout_url: string;
      provider_order_id: string;
      provider_api_version: string;
    };
  };
}

export class Checkout implements ICheckout {
  id: string;
  type: string;
  page_info: null;
  data: {
    payment_intent_id: string;
    account_id: string;
    platform_account_id: string;
    payment_amount: string;
    payment_client_id: string;
    payment_description: string;
    payment_methods: {
      id: string;
      status: string;
      invalid_reason: null;
      name: string;
      brand: string;
      acct_last_four: string;
      month: string;
      year: string;
      address_line1_check: string;
      address_postal_code_check: string;
      bin_details: null;
    }[];
    payment_method_group_id: string;
    payment_settings: {
      ach_payments: boolean;
      bnpl_payments: boolean;
      credit_card_payments: boolean;
    };
    bnpl?: {
      provider: string;
      provider_client_id: string;
      provider_mode: string;
      provider_checkout_url: string;
      provider_order_id: string;
      provider_api_version: string;
    };
  };

  constructor(data: ICheckout) {
    Object.assign(this, data);
  }
}
