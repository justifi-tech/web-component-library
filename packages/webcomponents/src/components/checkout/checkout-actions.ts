import { CheckoutService } from '../../api/services/checkout.service';
import { getErrorMessage } from '../../api/services/utils';

export const makeGetCheckout = ({ authToken, checkoutId, service }) =>
  async ({ onSuccess, onError }) => {
    try {
      const response = await (service as CheckoutService).fetchCheckout(authToken, checkoutId);

      if (!response.error) {
        const checkout = response.data;
        onSuccess({ checkout });
      } else {
        const responseError = getErrorMessage(response.error);
        return onError(responseError);
      }
    } catch (error) {
      return onError(error.message || error);
    }
  };

export const makePay = ({ authToken, checkoutId, service }) =>
  async ({ paymentMethodToken, onSuccess, onError }) => {
    try {
      const response = await (service as CheckoutService).pay(authToken, checkoutId, paymentMethodToken);

      if (!response.error) {
        const checkout = response.data;
        onSuccess({ checkout });
      } else {
        const responseError = getErrorMessage(response.error);
        return onError(responseError);
      }
    } catch (error) {
      return onError(error.message || error);
    }
  };
