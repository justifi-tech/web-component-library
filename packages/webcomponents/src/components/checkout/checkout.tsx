import { Component, Prop, h, State, Watch, Event, EventEmitter, Method, Listen } from '@stencil/core';
import { makeGetCheckout } from '../../actions/checkout/checkout-actions';
import { CheckoutService } from '../../api/services/checkout.service';
import JustifiAnalytics from '../../api/Analytics';
import { BillingFormFields } from './billing-form/billing-form-schema';
import { Checkout as CheckoutConstructor, ICheckout, ILoadedEventResponse } from '../../api';
import { checkPkgVersion } from '../../utils/check-pkg-version';
import { ComponentErrorEvent, ComponentSubmitEvent } from '../../api/ComponentEvents';
import { checkoutStore } from '../../store/checkout.store';
import { checkoutSummary } from '../../styles/parts';
import { insuranceValues, insuranceValuesOn } from '../insurance/insurance-state';
import { PaymentMethodTypes } from '../../api/Payment';
import { PaymentMethodOption } from './payment-method-option-utils';
import { StyledHost } from '../../ui-components';

const PaymentMethodTypeLabels = {
  bankAccount: 'New bank account',
  card: 'New credit or debit card',
};

@Component({
  tag: 'justifi-checkout',
})
export class Checkout {
  analytics: JustifiAnalytics;
  modularCheckoutRef?: HTMLJustifiModularCheckoutElement;
  selectedPaymentMethodOptionRef?: HTMLJustifiNewPaymentMethodElement | HTMLJustifiSavedPaymentMethodElement | HTMLJustifiSezzlePaymentMethodElement;

  @State() checkout: ICheckout;
  @State() complete: Function;
  @State() errorMessage: string = '';
  @State() insuranceToggled: boolean = false;
  @State() isSubmitting: boolean = false; // This is used to prevent multiple submissions and is different from loading state
  @State() paymentMethodOptions: PaymentMethodOption[] = [];
  @State() renderState: 'loading' | 'error' | 'success' = 'loading';
  @State() savePaymentMethod: boolean = false;
  @State() serverError: string;

  @Prop() authToken: string;
  @Prop() checkoutId: string;
  @Prop() disableBankAccount?: boolean = false;
  @Prop() disableBnpl?: boolean = false;
  @Prop() disableCreditCard?: boolean = false;
  @Prop() disablePaymentMethodGroup?: boolean = false;
  @Prop() hideBankAccountBillingForm?: boolean = false;
  @Prop() hideCardBillingForm?: boolean = false;

  @Watch('authToken')
  @Watch('checkoutId')
  @Watch('disableCreditCard')
  @Watch('disableBankAccount')
  @Watch('disableBnpl')
  @Watch('disablePaymentMethodGroup')
  propChanged() {
    checkoutStore.authToken = this.authToken;
    checkoutStore.checkoutId = this.checkoutId;
    checkoutStore.disableCreditCard = this.disableCreditCard;
    checkoutStore.disableBankAccount = this.disableBankAccount;
    checkoutStore.disableBnpl = this.disableBnpl;
    checkoutStore.disablePaymentMethodGroup = this.disablePaymentMethodGroup;
    this.fetchData();
  }

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentErrorEvent>;
  @Event({ eventName: 'loaded' }) loadedEvent: EventEmitter<ILoadedEventResponse>;
  @Event({ eventName: 'submit-event' }) submitEvent: EventEmitter<ComponentSubmitEvent>;

  connectedCallback() {
    if (this.authToken && this.checkoutId) {
      checkoutStore.checkoutId = this.checkoutId;
      checkoutStore.authToken = this.authToken;
      checkoutStore.disableCreditCard = this.disableCreditCard;
      checkoutStore.disableBankAccount = this.disableBankAccount;
      checkoutStore.disableBnpl = this.disableBnpl;
      checkoutStore.disablePaymentMethodGroup = this.disablePaymentMethodGroup;
    }
  }

  componentWillLoad() {
    checkPkgVersion();
    this.analytics = new JustifiAnalytics(this);

    this.fetchData();

    // Refresh the checkout data when insurance is added or removed
    insuranceValuesOn('set', (key) => {
      const value = insuranceValues[key];
      if (value !== undefined) {
        this.insuranceToggled = value;
        this.fetchData();
      }
    });
  }

  disconnectedCallback() {
    this.analytics?.cleanup();
  }

  @Listen('checkboxChanged')
  savePaymentMethodChanged(event: CustomEvent<boolean>) {
    this.savePaymentMethod = event.detail;
  }

  @Listen('checkout-complete-event')
  checkoutComplete(event: CustomEvent<any>) {
    this.isSubmitting = false;
    this.submitEvent.emit({
      response: event.detail,
    });
  }

  @Listen('error-event')
  checkoutError(event: CustomEvent<any>) {
    this.isSubmitting = false;
    console.error('checkout error', event.detail);
  }

  @Listen('paymentMethodOptionSelected')
  paymentMethodOptionSelected(event: CustomEvent<PaymentMethodOption>) {
    checkoutStore.selectedPaymentMethod = event.detail.id;
  }

  @Method()
  async fillBillingForm(fields: BillingFormFields) {
    checkoutStore.billingFormFields = fields;
  }

  @Method()
  async validate(): Promise<{ isValid: boolean }> {
    return { isValid: await this.modularCheckoutRef?.validate() };
  }

  private fetchData = (): void => {
    this.renderState = 'loading';

    const actionsConfig = {
      authToken: this.authToken,
      checkoutId: this.checkoutId,
      service: new CheckoutService()
    };
    const getCheckout = makeGetCheckout(actionsConfig);

    getCheckout({
      onSuccess: ({ checkout }) => {
        this.checkout = new CheckoutConstructor(checkout);
        const { status } = this.checkout;
        this.loadedEvent.emit({ checkout_status: status });
        this.renderState = 'success';
      },
      onError: ({ error, code, severity }) => {
        this.serverError = error;
        this.renderState = 'error';
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
      }
    });
  };

  private async submit(_event) {
    this.isSubmitting = true;
    this.modularCheckoutRef.submitCheckout(checkoutStore.billingFormFields);
  }

  private get canSavePaymentMethod() {
    return checkoutStore.selectedPaymentMethod === PaymentMethodTypes.card || checkoutStore.selectedPaymentMethod === PaymentMethodTypes.bankAccount;
  }

  private get isLoading() {
    return this.renderState === 'loading';
  }

  private get showBillingForm() {
    return (checkoutStore.selectedPaymentMethod === PaymentMethodTypes.card && !this.hideCardBillingForm)
      || (checkoutStore.selectedPaymentMethod === PaymentMethodTypes.bankAccount && !this.hideBankAccountBillingForm);
  }

  private get showBillingFormSection() {
    return checkoutStore.selectedPaymentMethod === PaymentMethodTypes.card || checkoutStore.selectedPaymentMethod === PaymentMethodTypes.bankAccount;
  }

  private get showPaymentTypeHeader() {
    return !this.disableCreditCard && !this.disableBankAccount;
  }

  private get showPostalCodeForm() {
    return checkoutStore.selectedPaymentMethod === PaymentMethodTypes.card && this.hideCardBillingForm;
  }

  render() {
    return (
      <StyledHost>
        <justifi-modular-checkout
          ref={(el) => (this.modularCheckoutRef = el)}
          authToken={this.authToken}
          checkoutId={this.checkoutId}
          savePaymentMethod={this.canSavePaymentMethod && this.savePaymentMethod}
        >
          <div class="row gy-3 jfi-checkout-core">
            <div class="col-12" part={checkoutSummary}>
              <justifi-header text="Summary" level="h2" class="fs-5 fw-bold pb-3" />
              <section>
                {this.isLoading && (
                  <justifi-skeleton height="24px" />
                )}
                {!this.isLoading && <justifi-checkout-summary />}
              </section>
            </div>
            <div class="col-12 mt-4">
              <slot name="insurance"></slot>
            </div>
            <div class="col-12 mt-4">
              <justifi-header text="Payment" level="h2" class="fs-5 fw-bold pb-3" />
              {this.showPaymentTypeHeader && (
                <justifi-header text="Select payment type" level="h3" class="fs-6 fw-bold lh-lg" />
              )}
              <div class="d-flex flex-column">
                <section>
                  {/* For now, just return nothing to avoid breaking, but we can decide to show an error message here */}
                  {/* <div style={{ color: 'red' }}>Error: {this.serverError}</div>; */}
                  {this.isLoading && (
                    <justifi-skeleton height="300px" />
                  )}
                  {!this.isLoading && (
                    <div>
                      <justifi-saved-payment-methods />
                      <justifi-sezzle-payment-method />
                      {!this.disableCreditCard && (
                        <div>
                          <payment-method-option
                            paymentMethodOptionId={PaymentMethodTypes.card}
                            isSelected={checkoutStore.selectedPaymentMethod === PaymentMethodTypes.card}
                            clickHandler={() => { checkoutStore.selectedPaymentMethod = PaymentMethodTypes.card }}
                            radioButtonHidden={this.disableCreditCard}
                            label={PaymentMethodTypeLabels[PaymentMethodTypes.card]}
                          />
                          {checkoutStore.selectedPaymentMethod === PaymentMethodTypes.card && (
                            <div class="mt-4 mb-4">
                              <justifi-card-form />
                            </div>
                          )}
                        </div>
                      )}
                      {!this.disableBankAccount && (
                        <div>
                          <payment-method-option
                            paymentMethodOptionId={PaymentMethodTypes.bankAccount}
                            isSelected={checkoutStore.selectedPaymentMethod === PaymentMethodTypes.bankAccount}
                            clickHandler={() => { checkoutStore.selectedPaymentMethod = PaymentMethodTypes.bankAccount }}
                            radioButtonHidden={this.disableBankAccount}
                            label={PaymentMethodTypeLabels[PaymentMethodTypes.bankAccount]}
                          />
                          {checkoutStore.selectedPaymentMethod === PaymentMethodTypes.bankAccount && (
                            <div class="mt-4 mb-4">
                              <justifi-bank-account-form />
                            </div>
                          )}
                        </div>
                      )}
                    </div>
                  )}
                </section>
              </div>
            </div>
            {this.showBillingFormSection && (
              <div class="col-12 mt-4">
                {this.showBillingForm && (
                  <justifi-header text="Billing Address" level="h2" class="fs-5 fw-bold pb-3" />
                )}
                {this.showPostalCodeForm && (
                  <justifi-postal-code-form />
                )}
                {this.showBillingForm && (
                  <justifi-billing-information-form />
                )}
              </div>
            )}
            <div class="col-12">
              {this.canSavePaymentMethod && (
                <justifi-save-new-payment-method />
              )}
            </div>
            <div class="mt-4">
              <justifi-button
                text="Pay"
                type="submit"
                variant="primary"
                clickHandler={(e) => this.submit(e)}
                disabled={this.isLoading || this.isSubmitting}
                isLoading={this.isLoading || this.isSubmitting}
                customStyle={{ width: '100%', textAlign: "center" }}
              />
            </div>
          </div>
        </justifi-modular-checkout>
      </StyledHost>
    );
  }
}
