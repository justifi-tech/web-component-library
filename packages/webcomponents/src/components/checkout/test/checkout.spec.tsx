import { newSpecPage } from '@stencil/core/testing';
import { Checkout } from '../checkout';
import { CheckoutCore } from '../checkout-core';
import JustifiAnalytics from '../../../api/Analytics';

beforeEach(() => {
  // Bypass Analytics to avoid errors. Analytics attaches events listeners to HTML elements
  // which are not available in Jest/node environment
  // @ts-ignore
  JustifiAnalytics.prototype.trackCustomEvents = jest.fn();
});

describe('justifi-checkout', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [Checkout, CheckoutCore],
      html: '<justifi-checkout></justifi-checkout>',
    });

    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });
});
