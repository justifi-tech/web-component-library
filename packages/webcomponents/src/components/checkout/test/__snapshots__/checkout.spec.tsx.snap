// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`justifi-checkout renders loading 1`] = `
<justifi-checkout class="body" style="display: block;">
  <!---->
  <style></style>
  <justifi-modular-checkout>
    <div class="gy-3 jfi-checkout-core row">
      <div class="col-12" part="checkout-summary">
        <justifi-header class="fs-5 fw-bold pb-3" level="h2" text="Summary"></justifi-header>
        <section>
          <justifi-checkout-summary></justifi-checkout-summary>
        </section>
      </div>
      <div class="col-12 mt-4"></div>
      <div class="col-12 mt-4">
        <justifi-header class="fs-5 fw-bold pb-3" level="h2" text="Payment"></justifi-header>
        <justifi-header class="fs-6 fw-bold lh-lg" level="h3" text="Select payment type"></justifi-header>
        <div class="d-flex flex-column">
          <section>
            <div>
              <justifi-saved-payment-methods></justifi-saved-payment-methods>
              <justifi-tokenize-payment-method accountid="" hidesubmitbutton=""></justifi-tokenize-payment-method>
            </div>
          </section>
        </div>
      </div>
      <div class="mt-4">
        <justifi-button text="Pay" type="submit" variant="primary"></justifi-button>
      </div>
    </div>
  </justifi-modular-checkout>
</justifi-checkout>
`;
